geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/key/AbstractDumpRestoreIntegrationTest.java

```diff
@@ -0,0 +1,257 @@
+import java.util.Set;
+
+import io.lettuce.core.KeyValue;
+import io.lettuce.core.RestoreArgs;
+import io.lettuce.core.cluster.RedisClusterClient;
+import io.lettuce.core.cluster.api.sync.RedisAdvancedClusterCommands;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import redis.clients.jedis.HostAndPort;
+import redis.clients.jedis.JedisCluster;
+import redis.clients.jedis.Protocol;
+
+import org.apache.geode.internal.InternalDataSerializer;
+import org.apache.geode.internal.serialization.DataSerializableFixedID;
+import org.apache.geode.redis.RedisIntegrationTest;
+import org.apache.geode.redis.internal.data.RedisHash;
+import org.apache.geode.redis.internal.data.RedisSet;
+import org.apache.geode.redis.internal.data.RedisSortedSet;
+import org.apache.geode.redis.internal.data.RedisString;
+
+public abstract class AbstractDumpRestoreIntegrationTest implements RedisIntegrationTest {
+
+  private RedisAdvancedClusterCommands<String, String> lettuce;
+  private JedisCluster jedis;
+  private static String STRING_VALUE;
+  private static byte[] RESTORE_BYTES;
+
+  @BeforeClass
+  public static void setupClass() {
+    InternalDataSerializer.getDSFIDSerializer().registerDSFID(
+        DataSerializableFixedID.REDIS_SET_ID,
+        RedisSet.class);
+    InternalDataSerializer.getDSFIDSerializer().registerDSFID(
+        DataSerializableFixedID.REDIS_STRING_ID,
+        RedisString.class);
+    InternalDataSerializer.getDSFIDSerializer().registerDSFID(
+        DataSerializableFixedID.REDIS_HASH_ID,
+        RedisHash.class);
+    InternalDataSerializer.getDSFIDSerializer().registerDSFID(
+        DataSerializableFixedID.REDIS_SORTED_SET_ID,
+        RedisSortedSet.class);
+  }
+
+  @Before
+  public void setup() {
+    jedis = new JedisCluster(new HostAndPort(BIND_ADDRESS, getPort()), REDIS_CLIENT_TIMEOUT);
+
+    RedisClusterClient client =
+        RedisClusterClient.create(String.format("redis://%s:%d", BIND_ADDRESS, getPort()));
+    lettuce = client.connect().sync();
+
+    STRING_VALUE = "It's a mad, mad, mad, mad, mad world";
+    lettuce.set("set-dump-value", STRING_VALUE);
+    RESTORE_BYTES = lettuce.dump("set-dump-value");
+  }
+
+  @After
+  public void tearDown() {
+    flushAll();
+  }
+
+  @Test
+  public void dumpTakesExactlyOneArgument() {
+    assertExactNumberOfArgs(jedis, Protocol.Command.DUMP, 1);
+  }
+
+  @Test
+  public void restoreErrorsWithUnknownOption() {
+    assertThatThrownBy(
+        () -> jedis.sendCommand("key", Protocol.Command.RESTORE, "key", "0", "", "FOO"))
+            .hasMessageContaining("ERR syntax error");
+  }
+
+  @Test
+  public void restoreFails_whenKeyAlreadyExists() {
+    lettuce.set("restored", "already exists");
+
+    assertThatThrownBy(() -> lettuce.restore("restored", 0, RESTORE_BYTES))
+        .hasMessage("BUSYKEY Target key name already exists.");
+  }
+
+  @Test
+  public void restoreFails_whenTTLisNegative() {
```

Comment:
Could we also add a test for when TTL is not an integer?

Reply:
Done.